quicklens/src/main/scala-2/com/softwaremill/quicklens/QuicklensMacros.scala

```diff
@@ -251,7 +251,7 @@ object QuicklensMacros {
         case FunctorPathElement(functor, method, xargs @ _*) :: tail =>
           val newRootPathEl = TermName(c.freshName())
           // combine the selected path with variable args
-          val args = generateSelects(newRootPathEl, tail) :: xargs.toList
+          val args = generateSelects(newRootPathEl, tail) :: xargs.toList.map(c.untypecheck(_))
```

Comment:
@KacperFKorban do you know any resource where I could read why is this necessary?

Reply:
I can't speak about the "Scala 2 macros" side of things, since this is the first piece of Scala 2 macro code that I've ever written :smile: But from what I understand, this change should essentially change the typed trees to untyped trees. Which in this case would fix owners of the argument closures.

When it comes to the issue itself, the problem is that if we have a snippet like this:

```scala
case class Foo(field: String)

def doStuff(): Unit = {
  val lang2 = "hey"

  Seq(Foo("asdf"))
    .modify(_.eachWhere(_.field != lang2).field)
    .setTo(lang2)
}
```

The macro will rewrite it to sth like:

```scala
def doStuff(): Unit = {
  val lang2: String = "hey"
  val obj = Seq(Foo("asdf"))
  com.softwaremill.quicklens.PathModify.apply[Seq[Foo], String](
    obj,
    { (mod: String => String) =>
      summon[QuicklensFunctor[Seq]].eachWhere[Foo](
        obj, 
        { (foo: Foo) => foo.copy(field = mod(foo.field)) }, 
        (_.field != lang2) // This is the problem
      )
    }
  ).setTo(lang2)
}
```

The function `(_.field != lang2)` was given as an argument and was used to construct the new 'result' expression, but we didn't correct its owner. The correct owner for it should be the `{ (mod: String => String) => ... }` lambda, but it stayed as the old owner i.e. `(_.eachWhere(_.field != lang2).field)`. This can then fail in a lot of places in the compiler, probably mostly when lifting nested functions and/or generating bytecode.

In Scala 3, this is checked under `-Ycheck:all` (might as well sprinkle `-Xcheck-macros` on top).

Reply:
I see, thanks for the explanation, it helped a lot!